### PR TITLE
[NOX-13] refactor plugin loading functionality

### PR DESCRIPTION
**Pull request by rafalmaziejuk**
Signed-off-by: rafalmaziejuk <rafalmaziejuk@gmail.com>

### CODE REVIEW

File: src/plugins/plugin.h [hunk 1]
```diff
@@ -21,17 +21,27 @@ class Plugin {
...
-    ConvertibleProcedureAddress getFunction(std::string_view name) const {
+    [[nodiscard]] static Plugin *load(std::string_view name);
+    [[nodiscard]] static std::string createFilename(std::string_view name, bool withPrefix, bool withPostfix);
+    [[nodiscard]] static std::string createFilenameWithExtension(std::string_view name, std::string_view extension);
+
+    [[nodiscard]] ConvertibleProcedureAddress operator[](std::string_view name) const {
         return {getProcedureAddress(name)};
     }
 
+    std::string getFilename() { return m_name; }
```
Comment (09pawel0898) [inline src/plugins/plugin.h:35]: Getter that is returning attribute copy should be const

File: src/plugins/plugin_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#include "plugins/plugin_manager.h"
+
+namespace NOX {
+
+PluginHandle PluginManager::registerPlugin(Plugin *plugin) {
+    NOX_ASSERT(plugin == nullptr);
+    NOX_ASSERT(isPluginRegistered(plugin->getFilename()));
+
+    const auto handle = static_cast<PluginHandle>(m_plugins.size() + 1);
+    m_plugins.insert({handle, std::unique_ptr<Plugin>(plugin)});
```
Comment (09pawel0898) [inline src/plugins/plugin_manager.cpp:11]: Avoid constructing std::unique_ptr by passing raw pointer to manually allocated object with new (RendererPlugin::load), use std::make_unique to create object instead. Then utilize move semantics to move the Plugin object to m_plugins container at the end.

File: src/plugins/plugin.h (hunk 1 above)
Comment (rafalmaziejuk) [reply src/plugins/plugin.h:35]: Missed it. Done.

File: src/plugins/plugin_manager.cpp (hunk 2 above)
Comment (rafalmaziejuk) [reply src/plugins/plugin_manager.cpp]: Corrected.

**Pull request merged**